### PR TITLE
feature/79: Improved performance by splitting reading of big parquet files

### LINKED ISSUE #79: Improve performance by splitting big parquet files
Improve performance by splitting big parquet files

### PR DESCRIPTION
**Pull request by jakobbraun**
Closes #79

### CODE REVIEW

File: doc/changes/changes_3.0.0.md [hunk 1]
```diff
@@ -1,33 +1,23 @@
-# Virtual Schema for document data in files 3.0.0, released 2021-??-??
+# Virtual Schema for document data in files 3.0.0, released 2021-11-22
 
-Code name:
+Code name: Improved Parquet Loading Performance
+
+## Summary
+
+In this release we improved the performance for loading parquet files. The improvements will give you a speed-up when you load a few (<200) files.
```
Comment (morazow) [inline doc/changes/changes_3.0.0.md:7]: Maybe also change this to "fewer than available CPUs"?

File: src/main/java/com/exasol/adapter/document/documentfetcher/files/parquet/ParquetDocumentFetcher.java [hunk 2]
```diff
@@ -1,42 +1,55 @@
...
-        try {
-            final ParquetReader<Row> reader = RowParquetReader.builder(hadoopInputFile).build();
-            return new AfterAllCallbackIterator<>(new ParquetIterator(reader), () -> tryToCloseReader(reader));
-        } catch (final IOException exception) {
-            throw new IllegalStateException(
-                    ExaError.messageBuilder("E-VSDF-7").message("Failed to read parquet file.").toString(), exception);
+    public Iterator<DocumentNode> readDocuments(final FileSegment segment) {
+        final InputFile hadoopInputFile = SeekableInputStreamAdapter
+                .convert(segment.getFile().getRandomAccessInputStream());
+        final ParquetFileSplitter splitter = new ParquetFileSplitter(hadoopInputFile, 16L * 1024);
```
Comment (morazow) [inline src/main/java/com/exasol/adapter/document/documentfetcher/files/parquet/ParquetDocumentFetcher.java:28]: Extract constant for chunk size

File: src/main/java/com/exasol/adapter/document/documentfetcher/files/segmentation/HashSegmentMatcher.java [hunk 3]
```diff
@@ -24,10 +27,14 @@ public HashSegmentMatcher(final HashSegmentDescription segmentDescription) {
...
 
     @Override
-    public boolean matches(final RemoteFile remoteFile) {
+    public List<FileSegment> getMatchingSegmentsFor(final RemoteFile remoteFile) {
         @java.lang.SuppressWarnings("squid:S2676") // abs hashcode is intended here
         final long hashNumber = Math.abs(remoteFile.getResourceName().hashCode());
         final int modulo = (int) (hashNumber % this.segmentDescription.getNumberOfSegments());
-        return modulo == this.segmentDescription.getSegmentId();
+        if (modulo == this.segmentDescription.getSegmentId()) {
+            return List.of(new FileSegment(remoteFile, new FileSegmentDescription(1, 0)));
```
Comment (morazow) [inline src/main/java/com/exasol/adapter/document/documentfetcher/files/segmentation/HashSegmentMatcher.java:35]: Use entire file constant also here

File: src/main/java/com/exasol/adapter/document/files/FileTypeSpecificDocumentFetcher.java [hunk 4]
```diff
@@ -21,5 +21,15 @@
      * @param remoteFile stream of the files contents with additional description for logging
      * @return read document nodes
      */
-    public Iterator<DocumentNode> readDocuments(RemoteFile remoteFile);
+    public Iterator<DocumentNode> readDocuments(FileSegment remoteFile);
+
+    /**
+     * Get if this document fetcher can read files partially.
+     * <p>
+     * Supporting this feature improves performance when loading fewer files then available CPU cores in the cluster.
```
Comment (morazow) [inline src/main/java/com/exasol/adapter/document/files/FileTypeSpecificDocumentFetcher.java:29]:
```suggestion
     * Supporting this feature improves performance when loading fewer files than available CPU cores in the cluster.
```

File: src/main/java/com/exasol/adapter/document/files/FilesDocumentFetcherFactory.java [hunk 5]
```diff
@@ -66,27 +65,48 @@ private StringFilter addPrefixToFilterToAvoidInjection(final StringFilter source
...
+        if (files.size() < numberOfSegments && fileSplittingIsSupported) {
+            final int factor = (int) Math.ceil((double) numberOfSegments / files.size());
+            for (final RemoteFile file : files) {
+                for (int splitCounter = 0; splitCounter < factor; splitCounter++) {
+                    splitFiles.add(new FileSegment(file, new FileSegmentDescription(factor, splitCounter)));
+                }
+            }
+        } else {
+            for (final RemoteFile file : files) {
+                splitFiles.add(new FileSegment(file, new FileSegmentDescription(1, 0)));
```
Comment (morazow) [inline src/main/java/com/exasol/adapter/document/files/FilesDocumentFetcherFactory.java:96]: Here also use ENTIRE_FILE constant

**Review by morazow — CHANGES_REQUESTED** [review]

File: doc/changes/changes_3.0.0.md (hunk 1 above)
Comment (jakobbraun) [reply doc/changes/changes_3.0.0.md:7]: 200 is hard coded here. If it's fewer than 200 the files are distributed explicitly otherwise by hash binning.

Comment (morazow) [reply doc/changes/changes_3.0.0.md:7]: Ahh yes, this for the other improvement.

**Review by morazow — APPROVED** [review]

**Pull request merged**